content/en/blog/_posts/2021-07-16-read-write-once-pod-access-mode-alpha.md

```diff
@@ -0,0 +1,269 @@
+In other cases maybe everyone is allowed to read data but not write it.
+For highly sensitive data, maybe only one user is allowed to read and write data but nobody else.
+
+In the world of Kubernetes, [access modes](https://kubernetes.io/docs/concepts/storage/persistent-volumes/#access-modes) are the way we model how storage is consumed.
+These access modes are a part of the spec for PersistentVolumes (PVs) and PersistentVolumeClaims (PVCs).
+
+```yaml
+kind: PersistentVolumeClaim
+apiVersion: v1
+metadata:
+  name: shared-cache
+spec:
+  accessModes:
+  - ReadWriteMany # Allow many pods to access shared-cache simultaneously.
+  resources:
+    requests:
+      storage: 1Gi
+```
+
+In Kubernetes, we currently have three access modes for PVs and PVCs:
+
+- ReadWriteOnce -- the volume can be mounted as read-write by a single node
+- ReadOnlyMany -- the volume can be mounted read-only by many nodes
+- ReadWriteMany -- the volume can be mounted as read-write by many nodes
+
+These access modes are enforced by Kubernetes components like the `kube-controller-manager` and `kubelet` to ensure only certain pods are allowed to access a given PersistentVolume.
+This allows users to follow the principle of least privilege to mitigate the impact of bad actors.
+
+## What is this new access mode and how does it work?
+
+In Kubernetes v1.22, a fourth access mode for PVs and PVCs is introduced:
+
+- ReadWriteOncePod -- the volume can be mounted as read-write by a single pod
+
+When a user creates a pod with a PVC that has the ReadWriteOncePod access mode, Kubernetes will ensure that pod is the only pod in the cluster that can read and write to it.
+
+If you create another pod that references the same PVC with this access mode, the pod will fail to start like so:
+
+```
+Events:
+  Type     Reason            Age   From               Message
+  ----     ------            ----  ----               -------
+  Warning  FailedScheduling  1s    default-scheduler  0/1 nodes are available: 1 node has pod using PersistentVolumeClaim with the same name and ReadWriteOncePod access mode.
+```
+
+### How is this different than the ReadWriteOnce access mode?
+
+The ReadWriteOnce access mode restricts volume access to a single *node*, which means it is possible for multiple pods on the same node to read from and write to the same volume.
+
+Could this cause problems?
+Potentially yes!
+If you are running workloads that access sensitive data, other pods on that node can access it (if updated to mount the volume).
+Unfortunately, the only way to work around this is by running a single pod on the node, which is an inefficient use of compute resources in your cluster.
+
+With ReadWriteOncePod these restrictions go away.
+You don't have to worry about neighboring pods potentially accessing your volumes.
+Just set the access mode on your PVC, and Kubernetes will guarantee only a single pod has access.
+
+## How do I use it?
```

Comment:
Note that this is only supported for CSI plugins.

As an aside, what happens if you enable this on an intree plugin?

Reply:
Done. If you enable this on an intree plugin it should work, we just have to add ReadWriteOncePod to [the list of supported access modes](https://github.com/kubernetes/kubernetes/blob/v1.22.0-rc.0/pkg/volume/gcepd/gce_pd.go#L113-L118).

Reply:
What was the reason why we decided not to support it for in-tree volumes if nothing stops a user from using it and it works?

Reply:
See https://github.com/kubernetes/website/pull/29059#issuecomment-889289648

Reply:
If we wanted to open this up to in-tree drivers we could do this for beta graduation.

Reply:
Synced with @Jiawei0227 , I did not know the API fields for in-tree drivers were not being removed. If customers want to rely on the legacy API fields + CSI migration and we feel it will help target more users we can add support here

Reply:
Yeah I think given that the enforcement is all done in Kubernetes, the plugin is not involved, and we don't error/reject if someone uses it with an in-tree type, it is better to support it for in-tree types as well so it's consistent across environments.

Reply:
If the implementation of ReadWriteOncePod is done in kubelet or KCM not in each intree plugins code, I think it is fine to support it for intree as well. 

What I was thinking is if we need to edit intree plugin codes that will be not worth it as most of the PersistentVolume plugin types will be either deprecated or removed.